# TV Script Generation

In this project, you'll generate your own [Seinfeld](https://en.wikipedia.org/wiki/Seinfeld) TV scripts using RNNs.  You'll be using part of the [Seinfeld dataset](https://www.kaggle.com/thec03u5/seinfeld-chronicles#scripts.csv) of scripts from 9 seasons.  The Neural Network you'll build will generate a new ,"fake" TV script, based on patterns it recognizes in this training data.

## Get the Data

The data is already provided for you in `./data/Seinfeld_Scripts.txt` and you're encouraged to open that file and look at the text. 
>* As a first step, we'll load in this data and look at some samples. 
* Then, you'll be tasked with defining and training an RNN to generate a new script!

In [1]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# load in data
import helper
data_dir = './data/Seinfeld_Scripts.txt'
text = helper.load_data(data_dir)

## Explore the Data
Play around with `view_line_range` to view different parts of the data. This will give you a sense of the data you'll be working with. You can see, for example, that it is all lowercase text, and each new line of dialogue is separated by a newline character `\n`.

In [2]:
view_line_range = (20, 50)

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
import numpy as np

print('Dataset Stats')
print('Roughly the number of unique words: {}'.format(len({word: None for word in text.split()})))

lines = text.split('\n')
print('Number of lines: {}'.format(len(lines)))
word_count_line = [len(line.split()) for line in lines]
print('Average number of words in each line: {}'.format(np.average(word_count_line)))

print()
print('The lines {} to {}:'.format(*view_line_range))
print('\n'.join(text.split('\n')[view_line_range[0]:view_line_range[1]]))

Dataset Stats
Roughly the number of unique words: 46367
Number of lines: 109233
Average number of words in each line: 5.544240293684143

The lines 20 to 50:
claire: its missing, i have to do it in my head decaf left, regular right, decaf left, regular right...its very challenging work. 

jerry: can you relax, its a cup of coffee. claire is a professional waitress. 

claire: trust me george. no one has any interest in seeing you on caffeine. 

george: how come youre not doing the second show tomorrow? 

jerry: well, theres this uh, woman might be coming in. 

george: wait a second, wait a second, what coming in, what woman is coming in? 

jerry: i told you about laura, the girl i met in michigan? 

george: no, you didnt! 

jerry: i thought i told you about it, yes, she teaches political science? i met her the night i did the show in lansing... 

george: ha. 

jerry: (looks in the creamer) theres no milk in here, what... 

george: wait wait wait, what is she... (takes the milk can from j

---
## Implement Pre-processing Functions
The first thing to do to any dataset is pre-processing.  Implement the following pre-processing functions below:
- Lookup Table
- Tokenize Punctuation

### Lookup Table
To create a word embedding, you first need to transform the words to ids.  In this function, create two dictionaries:
- Dictionary to go from the words to an id, we'll call `vocab_to_int`
- Dictionary to go from the id to word, we'll call `int_to_vocab`

Return these dictionaries in the following **tuple** `(vocab_to_int, int_to_vocab)`

In [3]:
import problem_unittests as tests
from collections import Counter

def create_lookup_tables(text):
    """
    Create lookup tables for vocabulary
    :param text: The text of tv scripts split into words
    :return: A tuple of dicts (vocab_to_int, int_to_vocab)
    """
    # TODO: Implement Function
    word_counts = Counter(text)
    sorted_vocab = sorted(word_counts, key=word_counts.get, reverse=True)
    int_to_vocab = { i:word for i, word in enumerate(sorted_vocab)}
    vocab_to_int = { word:i for i, word in int_to_vocab.items() }
    
    # return tuple
    return vocab_to_int, int_to_vocab


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_create_lookup_tables(create_lookup_tables)

Tests Passed


### Tokenize Punctuation
We'll be splitting the script into a word array using spaces as delimiters.  However, punctuations like periods and exclamation marks can create multiple ids for the same word. For example, "bye" and "bye!" would generate two different word ids.

Implement the function `token_lookup` to return a dict that will be used to tokenize symbols like "!" into "||Exclamation_Mark||".  Create a dictionary for the following symbols where the symbol is the key and value is the token:
- Period ( **.** )
- Comma ( **,** )
- Quotation Mark ( **"** )
- Semicolon ( **;** )
- Exclamation mark ( **!** )
- Question mark ( **?** )
- Left Parentheses ( **(** )
- Right Parentheses ( **)** )
- Dash ( **-** )
- Return ( **\n** )

This dictionary will be used to tokenize the symbols and add the delimiter (space) around it.  This separates each symbols as its own word, making it easier for the neural network to predict the next word. Make sure you don't use a value that could be confused as a word; for example, instead of using the value "dash", try using something like "||dash||".

In [4]:
def token_lookup():
    """
    Generate a dict to turn punctuation into a token.
    :return: Tokenized dictionary where the key is the punctuation and the value is the token
    """
    # TODO: Implement Function
        
    token_dict = dict()
    token_dict['.'] = "||PERIOD||"
    token_dict[','] = "||COMMA||"
    token_dict['"'] = "||QUOTATION_MARK||"
    token_dict[';'] = "||SEMI_COLON||"
    token_dict['!'] = "||EXCLAMATION_MARK||"
    token_dict['?'] = "||QUESTION_MARK||"
    token_dict['('] = "||LEFT_PARENTHESIS||"
    token_dict[')'] = "||RIGHT_PARENTHESIS||"
    token_dict['-'] = "||DASH||"
    token_dict['\n'] = "||RETURN||"
        
    return token_dict

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_tokenize(token_lookup)

Tests Passed


## Pre-process all the data and save it

Running the code cell below will pre-process all the data and save it to file. You're encouraged to lok at the code for `preprocess_and_save_data` in the `helpers.py` file to see what it's doing in detail, but you do not need to change this code.

In [5]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# pre-process training data
helper.preprocess_and_save_data(data_dir, token_lookup, create_lookup_tables)

# Check Point
This is your first checkpoint. If you ever decide to come back to this notebook or have to restart the notebook, you can start from here. The preprocessed data has been saved to disk.

In [6]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import helper
import problem_unittests as tests

int_text, vocab_to_int, int_to_vocab, token_dict = helper.load_preprocess()

## Build the Neural Network
In this section, you'll build the components necessary to build an RNN by implementing the RNN Module and forward and backpropagation functions.

### Check Access to GPU

In [7]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import torch

# Check for a GPU
train_on_gpu = torch.cuda.is_available()
if not train_on_gpu:
    print('No GPU found. Please use a GPU to train your neural network.')

## Input
Let's start with the preprocessed input data. We'll use [TensorDataset](http://pytorch.org/docs/master/data.html#torch.utils.data.TensorDataset) to provide a known format to our dataset; in combination with [DataLoader](http://pytorch.org/docs/master/data.html#torch.utils.data.DataLoader), it will handle batching, shuffling, and other dataset iteration functions.

You can create data with TensorDataset by passing in feature and target tensors. Then create a DataLoader as usual.
```
data = TensorDataset(feature_tensors, target_tensors)
data_loader = torch.utils.data.DataLoader(data, 
                                          batch_size=batch_size)
```

### Batching
Implement the `batch_data` function to batch `words` data into chunks of size `batch_size` using the `TensorDataset` and `DataLoader` classes.

>You can batch words using the DataLoader, but it will be up to you to create `feature_tensors` and `target_tensors` of the correct size and content for a given `sequence_length`.

For example, say we have these as input:
```
words = [1, 2, 3, 4, 5, 6, 7]
sequence_length = 4
```

Your first `feature_tensor` should contain the values:
```
[1, 2, 3, 4]
```
And the corresponding `target_tensor` should just be the next "word"/tokenized word value:
```
5
```
This should continue with the second `feature_tensor`, `target_tensor` being:
```
[2, 3, 4, 5]  # features
6             # target
```

In [8]:
from torch.utils.data import TensorDataset, DataLoader


def batch_data(words, sequence_length, batch_size):
    """
    Batch the neural network data using DataLoader
    :param words: The word ids of the TV scripts
    :param sequence_length: The sequence length of each batch
    :param batch_size: The size of each batch; the number of sequences in a batch
    :return: DataLoader with batched data
    """
    # TODO: Implement function
    n_batches = len(words)//batch_size
    
    words = words[:(n_batches*batch_size)]
    
    batch_x, batch_y = [], []
    
    for idx in range(0, len(words)-sequence_length):
        batch_x.append(words[idx:idx+sequence_length])
        batch_y.append(words[idx+sequence_length])
    
    data = TensorDataset(torch.from_numpy(np.asarray(batch_x)), torch.from_numpy(np.asarray(batch_y)))
    
    return DataLoader(data, shuffle=False, batch_size=batch_size)

# there is no test for this function, but you are encouraged to create
# print statements and tests of your own


### Test your dataloader 

You'll have to modify this code to test a batching function, but it should look fairly similar.

Below, we're generating some test text data and defining a dataloader using the function you defined, above. Then, we are getting some sample batch of inputs `sample_x` and targets `sample_y` from our dataloader.

Your code should return something like the following (likely in a different order, if you shuffled your data):

```
torch.Size([10, 5])
tensor([[ 28,  29,  30,  31,  32],
        [ 21,  22,  23,  24,  25],
        [ 17,  18,  19,  20,  21],
        [ 34,  35,  36,  37,  38],
        [ 11,  12,  13,  14,  15],
        [ 23,  24,  25,  26,  27],
        [  6,   7,   8,   9,  10],
        [ 38,  39,  40,  41,  42],
        [ 25,  26,  27,  28,  29],
        [  7,   8,   9,  10,  11]])

torch.Size([10])
tensor([ 33,  26,  22,  39,  16,  28,  11,  43,  30,  12])
```

### Sizes
Your sample_x should be of size `(batch_size, sequence_length)` or (10, 5) in this case and sample_y should just have one dimension: batch_size (10). 

### Values

You should also notice that the targets, sample_y, are the *next* value in the ordered test_text data. So, for an input sequence `[ 28,  29,  30,  31,  32]` that ends with the value `32`, the corresponding output should be `33`.

In [9]:
# test dataloader

test_text = range(50)
t_loader = batch_data(test_text, sequence_length=5, batch_size=10)

data_iter = iter(t_loader)
sample_x, sample_y = data_iter.next()

print(sample_x.shape)
print(sample_x)
print()
print(sample_y.shape)
print(sample_y)

torch.Size([10, 5])
tensor([[  0,   1,   2,   3,   4],
        [  1,   2,   3,   4,   5],
        [  2,   3,   4,   5,   6],
        [  3,   4,   5,   6,   7],
        [  4,   5,   6,   7,   8],
        [  5,   6,   7,   8,   9],
        [  6,   7,   8,   9,  10],
        [  7,   8,   9,  10,  11],
        [  8,   9,  10,  11,  12],
        [  9,  10,  11,  12,  13]])

torch.Size([10])
tensor([  5,   6,   7,   8,   9,  10,  11,  12,  13,  14])


---
## Build the Neural Network
Implement an RNN using PyTorch's [Module class](http://pytorch.org/docs/master/nn.html#torch.nn.Module). You may choose to use a GRU or an LSTM. To complete the RNN, you'll have to implement the following functions for the class:
 - `__init__` - The initialize function. 
 - `init_hidden` - The initialization function for an LSTM/GRU hidden state
 - `forward` - Forward propagation function.
 
The initialize function should create the layers of the neural network and save them to the class. The forward propagation function will use these layers to run forward propagation and generate an output and a hidden state.

**The output of this model should be the *last* batch of word scores** after a complete sequence has been processed. That is, for each input sequence of words, we only want to output the word scores for a single, most likely, next word.

### Hints

1. Make sure to stack the outputs of the lstm to pass to your fully-connected layer, you can do this with `lstm_output = lstm_output.contiguous().view(-1, self.hidden_dim)`
2. You can get the last batch of word scores by shaping the output of the final, fully-connected layer like so:

```
# reshape into (batch_size, seq_length, output_size)
output = output.view(batch_size, -1, self.output_size)
# get last batch
out = output[:, -1]
```

In [11]:
import torch.nn as nn
import problem_unittests as tests

class RNN(nn.Module):
    
    def __init__(self, vocab_size, output_size, embedding_dim, hidden_dim, n_layers, dropout=0.5):
        """
        Initialize the PyTorch RNN Module
        :param vocab_size: The number of input dimensions of the neural network (the size of the vocabulary)
        :param output_size: The number of output dimensions of the neural network
        :param embedding_dim: The size of embeddings, should you choose to use them        
        :param hidden_dim: The size of the hidden layer outputs
        :param dropout: dropout to add in between LSTM/GRU layers
        """
        super(RNN, self).__init__()
        # TODO: Implement function
        
        # set class variables
        self.embedding = nn.Embedding(vocab_size, embedding_dim)
        
        self.n_layers = n_layers
        self.hidden_dim = hidden_dim
        self.output_size = output_size
        
        # define model layers
        self.lstm = nn.LSTM(embedding_dim, hidden_dim, n_layers, dropout=dropout, batch_first=True)
        
        self.fc = nn.Linear(hidden_dim, output_size)
    
    
    def forward(self, nn_input, hidden):
        """
        Forward propagation of the neural network
        :param nn_input: The input to the neural network
        :param hidden: The hidden state        
        :return: Two Tensors, the output of the neural network and the latest hidden state
        """
        # TODO: Implement function   
        batch_size = nn_input.shape[0]
        embeddings = self.embedding(nn_input)
        lstm_output, hidden = self.lstm(embeddings, hidden)
        
        lstm_output = lstm_output.contiguous().view(-1, self.hidden_dim)
        
        output = self.fc(lstm_output)
        
        # reshape into (batch_size, seq_length, output_size)
        output = output.view(batch_size, -1, self.output_size)
        # get last batch
        output = output[:, -1]

        # return one batch of output word scores and the hidden state
        return output, hidden
    
    
    def init_hidden(self, batch_size):
        '''
        Initialize the hidden state of an LSTM/GRU
        :param batch_size: The batch_size of the hidden state
        :return: hidden state of dims (n_layers, batch_size, hidden_dim)
        '''
        # Implement function
        weight = next(self.parameters()).data
        
        if (train_on_gpu):
            hidden = (weight.new(self.n_layers, batch_size, self.hidden_dim).zero_().cuda(),
                      weight.new(self.n_layers, batch_size, self.hidden_dim).zero_().cuda())
        else:
            hidden = (weight.new(self.n_layers, batch_size, self.hidden_dim).zero_(),
                      weight.new(self.n_layers, batch_size, self.hidden_dim).zero_())
        
        return hidden

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_rnn(RNN, train_on_gpu)

Tests Passed


### Define forward and backpropagation

Use the RNN class you implemented to apply forward and back propagation. This function will be called, iteratively, in the training loop as follows:
```
loss = forward_back_prop(decoder, decoder_optimizer, criterion, inp, target)
```

And it should return the average loss over a batch and the hidden state returned by a call to `RNN(inp, hidden)`. Recall that you can get this loss by computing it, as usual, and calling `loss.item()`.

**If a GPU is available, you should move your data to that GPU device, here.**

In [12]:
def forward_back_prop(rnn, optimizer, criterion, inp, target, hidden):
    """
    Forward and backward propagation on the neural network
    :param decoder: The PyTorch Module that holds the neural network
    :param decoder_optimizer: The PyTorch optimizer for the neural network
    :param criterion: The PyTorch loss function
    :param inp: A batch of input to the neural network
    :param target: The target output for the batch of input
    :return: The loss and the latest hidden state Tensor
    """
    
    # TODO: Implement Function
    optimizer.zero_grad()
    
    # move data to GPU, if available
    if train_on_gpu:
        rnn, inp, target = rnn.cuda(), inp.cuda(), target.cuda()
    
    # Creating new variables for the hidden state
    h = tuple([ each.data for each in hidden ])
    
    output, h = rnn(inp, h)
    loss = criterion(output, target)
    
    # perform backpropagation and optimization
    loss.backward()
    nn.utils.clip_grad_norm_(rnn.parameters(), 5)
    optimizer.step()

    # return the loss over a batch and the hidden state produced by our model
    return loss.item(), h

# Note that these tests aren't completely extensive.
# they are here to act as general checks on the expected outputs of your functions
"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_forward_back_prop(RNN, forward_back_prop, train_on_gpu)

Tests Passed


## Neural Network Training

With the structure of the network complete and data ready to be fed in the neural network, it's time to train it.

### Train Loop

The training loop is implemented for you in the `train_decoder` function. This function will train the network over all the batches for the number of epochs given. The model progress will be shown every number of batches. This number is set with the `show_every_n_batches` parameter. You'll set this parameter along with other parameters in the next section.

In [13]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""

def train_rnn(rnn, batch_size, optimizer, criterion, n_epochs, show_every_n_batches=100):
    batch_losses = []
    
    rnn.train()

    print("Training for %d epoch(s)..." % n_epochs)
    for epoch_i in range(1, n_epochs + 1):
        
        # initialize hidden state
        hidden = rnn.init_hidden(batch_size)
        
        for batch_i, (inputs, labels) in enumerate(train_loader, 1):
            
            # make sure you iterate over completely full batches, only
            n_batches = len(train_loader.dataset)//batch_size
            if(batch_i > n_batches):
                break
            
            # forward, back prop
            loss, hidden = forward_back_prop(rnn, optimizer, criterion, inputs, labels, hidden)          
            # record loss
            batch_losses.append(loss)

            # printing loss stats
            if batch_i % show_every_n_batches == 0:
                print('Epoch: {:>4}/{:<4}  Loss: {}\n'.format(
                    epoch_i, n_epochs, np.average(batch_losses)))
                batch_losses = []

    # returns a trained rnn
    return rnn

### Hyperparameters

Set and train the neural network with the following parameters:
- Set `sequence_length` to the length of a sequence.
- Set `batch_size` to the batch size.
- Set `num_epochs` to the number of epochs to train for.
- Set `learning_rate` to the learning rate for an Adam optimizer.
- Set `vocab_size` to the number of uniqe tokens in our vocabulary.
- Set `output_size` to the desired size of the output.
- Set `embedding_dim` to the embedding dimension; smaller than the vocab_size.
- Set `hidden_dim` to the hidden dimension of your RNN.
- Set `n_layers` to the number of layers/cells in your RNN.
- Set `show_every_n_batches` to the number of batches at which the neural network should print progress.

If the network isn't getting the desired results, tweak these parameters and/or the layers in the `RNN` class.

In [14]:
# Data params
# Sequence Length
sequence_length = 10  # of words in a sequence
# Batch Size
batch_size = 128

# data loader - do not change
train_loader = batch_data(int_text, sequence_length, batch_size)

In [15]:
# Training parameters
# Number of Epochs
num_epochs = 10
# Learning Rate
learning_rate = 1e-3

# Model parameters
# Vocab size
vocab_size = len(vocab_to_int)
# Output size
output_size = vocab_size
# Embedding Dimension
embedding_dim = 200
# Hidden Dimension
hidden_dim = 250
# Number of RNN Layers
n_layers = 2

# Show stats for every n number of batches
show_every_n_batches = 2500


### Train
In the next cell, you'll train the neural network on the pre-processed data.  If you have a hard time getting a good loss, you may consider changing your hyperparameters. In general, you may get better results with larger hidden and n_layer dimensions, but larger models take a longer time to train. 
> **You should aim for a loss less than 3.5.** 

You should also experiment with different sequence lengths, which determine the size of the long range dependencies that a model can learn.

In [16]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""

# create model and move to gpu if available
rnn = RNN(vocab_size, output_size, embedding_dim, hidden_dim, n_layers, dropout=0.5)
if train_on_gpu:
    rnn.cuda()

# defining loss and optimization functions for training
optimizer = torch.optim.Adam(rnn.parameters(), lr=learning_rate)
criterion = nn.CrossEntropyLoss()

# training the model
trained_rnn = train_rnn(rnn, batch_size, optimizer, criterion, num_epochs, show_every_n_batches)

# saving the trained model
helper.save_model('./save/trained_rnn', trained_rnn)
print('Model Trained and Saved')

Training for 2 epoch(s)...
Epoch:    1/2     Loss: 5.483266196727753

Epoch:    1/2     Loss: 5.001698865890503

Epoch:    1/2     Loss: 4.835119313955307

Epoch:    1/2     Loss: 4.763312108755112

Epoch:    1/2     Loss: 4.687075222015381

Epoch:    1/2     Loss: 4.5987860288619995

Epoch:    1/2     Loss: 4.589814869880676

Epoch:    1/2     Loss: 4.5478133995532986

Epoch:    2/2     Loss: 4.50605746782131

Epoch:    2/2     Loss: 4.41265679359436

Epoch:    2/2     Loss: 4.40928323674202

Epoch:    2/2     Loss: 4.391344643831253

Epoch:    2/2     Loss: 4.411121933460236

Epoch:    2/2     Loss: 4.389279869318009

Epoch:    2/2     Loss: 4.364561497688293

Epoch:    2/2     Loss: 4.3517549700737

Model Trained and Saved


/home/ec2-user/anaconda3/lib/python3.7/site-packages/torch/serialization.py:251: UserWarning: Couldn't retrieve source code for container of type RNN. It won't be checked for correctness upon loading.
  "type " + obj.__name__ + ". It won't be checked "


In [14]:
# Data params
# Sequence Length
sequence_length = 50  # of words in a sequence
# Batch Size
batch_size = 128

# data loader - do not change
train_loader = batch_data(int_text, sequence_length, batch_size)

# Training parameters
# Number of Epochs
num_epochs = 5
# Learning Rate
learning_rate = 1e-3

# Model parameters
# Vocab size
vocab_size = len(vocab_to_int)
# Output size
output_size = vocab_size
# Embedding Dimension
embedding_dim = 200
# Hidden Dimension
hidden_dim = 250
# Number of RNN Layers
n_layers = 2

# Show stats for every n number of batches
show_every_n_batches = 1500

"""
DON'T MODIFY ANYTHING IN THIS CELL
"""

# create model and move to gpu if available
rnn = RNN(vocab_size, output_size, embedding_dim, hidden_dim, n_layers, dropout=0.5)
if train_on_gpu:
    rnn.cuda()

# defining loss and optimization functions for training
optimizer = torch.optim.Adam(rnn.parameters(), lr=learning_rate)
criterion = nn.CrossEntropyLoss()

# training the model
from workspace_utils import active_session
with active_session():
    trained_rnn = train_rnn(rnn, batch_size, optimizer, criterion, num_epochs, show_every_n_batches)

# saving the trained model
helper.save_model('./save/trained_rnn', trained_rnn)
print('Model Trained and Saved')

Training for 5 epoch(s)...
Epoch:    1/5     Loss: 5.178389035542806

Epoch:    1/5     Loss: 4.647089878559113

Epoch:    1/5     Loss: 4.4883689101537065

Epoch:    1/5     Loss: 4.416701443354289

Epoch:    2/5     Loss: 4.305928659614323

Epoch:    2/5     Loss: 4.242049318154653

Epoch:    2/5     Loss: 4.20496751499176

Epoch:    2/5     Loss: 4.196145922660827

Epoch:    3/5     Loss: 4.143069757934296

Epoch:    3/5     Loss: 4.109075900713603

Epoch:    3/5     Loss: 4.0964836138089495

Epoch:    3/5     Loss: 4.095508715311686

Epoch:    4/5     Loss: 4.0373095108957475

Epoch:    4/5     Loss: 4.032995078722636

Epoch:    4/5     Loss: 4.018283849398295

Epoch:    4/5     Loss: 4.0342941241264345

Epoch:    5/5     Loss: 3.9935848118091615

Epoch:    5/5     Loss: 3.971880710601807

Epoch:    5/5     Loss: 3.981141970316569

Epoch:    5/5     Loss: 3.980663827896118

Model Trained and Saved


/opt/conda/lib/python3.6/site-packages/torch/serialization.py:193: UserWarning: Couldn't retrieve source code for container of type RNN. It won't be checked for correctness upon loading.
  "type " + obj.__name__ + ". It won't be checked "


In [15]:
# Re-Training parameters
# Number of Epochs
num_epochs = 5
# Learning Rate
learning_rate = 1e-5

# training the model
from workspace_utils import active_session
with active_session():
    trained_rnn = train_rnn(rnn, batch_size, optimizer, criterion, num_epochs, show_every_n_batches)

# saving the trained model
helper.save_model('./save/trained_rnn', trained_rnn)
print('Model Trained and Saved')

Training for 5 epoch(s)...
Epoch:    1/5     Loss: 3.918714995702108

Epoch:    1/5     Loss: 3.920423555533091

Epoch:    1/5     Loss: 3.936611859003703

Epoch:    1/5     Loss: 3.960960993131002

Epoch:    2/5     Loss: 3.9041548083878967

Epoch:    2/5     Loss: 3.9050394859313964

Epoch:    2/5     Loss: 3.9100418181419374

Epoch:    2/5     Loss: 3.9058615531921386

Epoch:    3/5     Loss: 3.87681698735555

Epoch:    3/5     Loss: 3.887181424776713

Epoch:    4/5     Loss: 3.852043380246643

Epoch:    4/5     Loss: 3.852189881801605

Epoch:    4/5     Loss: 3.8558993360201517

Epoch:    4/5     Loss: 3.8631221129099527

Epoch:    5/5     Loss: 3.8254242522612354

Epoch:    5/5     Loss: 3.8415723021825157

Epoch:    5/5     Loss: 3.8519060818354287

Model Trained and Saved


/opt/conda/lib/python3.6/site-packages/torch/serialization.py:193: UserWarning: Couldn't retrieve source code for container of type RNN. It won't be checked for correctness upon loading.
  "type " + obj.__name__ + ". It won't be checked "


In [16]:
# Re-Training parameters
# Number of Epochs
num_epochs = 8
# Learning Rate
learning_rate = 1e-6

# training the model
from workspace_utils import active_session
with active_session():
    trained_rnn = train_rnn(rnn, batch_size, optimizer, criterion, num_epochs, show_every_n_batches)

# saving the trained model
helper.save_model('./save/trained_rnn', trained_rnn)
print('Model Trained and Saved')

Training for 8 epoch(s)...
Epoch:    1/8     Loss: 3.7928017174402875

Epoch:    1/8     Loss: 3.804341961701711

Epoch:    1/8     Loss: 3.804695385138194

Epoch:    1/8     Loss: 3.817438975652059

Epoch:    2/8     Loss: 3.7952040284446134

Epoch:    2/8     Loss: 3.7806702213287355

Epoch:    2/8     Loss: 3.8082194571495056

Epoch:    2/8     Loss: 3.801771854718526

Epoch:    3/8     Loss: 3.771344301602265

Epoch:    3/8     Loss: 3.7690249369939166

Epoch:    3/8     Loss: 3.7789699767430625

Epoch:    3/8     Loss: 3.792053918679555

Epoch:    4/8     Loss: 3.7661322411248617

Epoch:    4/8     Loss: 3.7461767597198485

Epoch:    4/8     Loss: 3.7654677993456525

Epoch:    4/8     Loss: 3.7697931264241538

Epoch:    5/8     Loss: 3.7468061508476223

Epoch:    5/8     Loss: 3.729994402249654

Epoch:    5/8     Loss: 3.7532857073148094

Epoch:    5/8     Loss: 3.761983794371287

Epoch:    6/8     Loss: 3.7320469194842922

Epoch:    6/8     Loss: 3.7226834707260132

Epoch:    6/8

/opt/conda/lib/python3.6/site-packages/torch/serialization.py:193: UserWarning: Couldn't retrieve source code for container of type RNN. It won't be checked for correctness upon loading.
  "type " + obj.__name__ + ". It won't be checked "


In [22]:
# Re-Training parameters
# Number of Epochs
num_epochs = 15
# Learning Rate
learning_rate = 1e-3
num_epochs = 15
sequence_length = 50
batch_size = 128
train_loader = batch_data(int_text, sequence_length, batch_size)
embedding_dim = 200
hidden_dim = 250
n_layers = 2
show_every_n_batches = 2000

# training the model
from workspace_utils import active_session

rnn = helper.load_model('./save/trained_rnn')
if train_on_gpu:
    rnn.cuda()

# defining loss and optimization functions for training
optimizer = torch.optim.Adam(rnn.parameters(), lr=learning_rate)
criterion = nn.CrossEntropyLoss()

with active_session():
    trained_rnn = train_rnn(rnn, batch_size, optimizer, criterion, num_epochs, show_every_n_batches)

# saving the trained model
helper.save_model('./save/trained_rnn', trained_rnn)
print('Model Trained and Saved')


Training for 15 epoch(s)...
Epoch:    1/15    Loss: 5.704716923952103

Epoch:    1/15    Loss: 4.952613417267799

Epoch:    1/15    Loss: 4.752679703474045

Epoch:    2/15    Loss: 4.587937940492465

Epoch:    2/15    Loss: 4.480507374286652

Epoch:    2/15    Loss: 4.444903503537178

Epoch:    3/15    Loss: 4.366542076492438

Epoch:    3/15    Loss: 4.300349650740624

Epoch:    3/15    Loss: 4.299374842524529

Epoch:    4/15    Loss: 4.242838933766794

Epoch:    4/15    Loss: 4.217219951748848

Epoch:    4/15    Loss: 4.203390802979469

Epoch:    5/15    Loss: 4.1584571258729275

Epoch:    5/15    Loss: 4.145925901412964

Epoch:    5/15    Loss: 4.154335049271584

Epoch:    6/15    Loss: 4.11018447337742

Epoch:    6/15    Loss: 4.0935953426361085

Epoch:    6/15    Loss: 4.095695605993271

Epoch:    7/15    Loss: 4.061083738366347

Epoch:    7/15    Loss: 4.049079134345055

Epoch:    7/15    Loss: 4.069152877092361

Epoch:    8/15    Loss: 4.025918975074616

Epoch:    8/15    Loss: 4

/opt/conda/lib/python3.6/site-packages/torch/serialization.py:193: UserWarning: Couldn't retrieve source code for container of type RNN. It won't be checked for correctness upon loading.
  "type " + obj.__name__ + ". It won't be checked "


In [14]:
# Re-training parameters
# Number of Epochs
num_epochs = 20
learning_rate = 1e-3
vocab_size = len(vocab_to_int)
output_size = vocab_size
embedding_dim = 200
hidden_dim = 250
n_layers = 2
show_every_n_batches = 2000


# training the model
from workspace_utils import active_session

rnn = RNN(vocab_size, output_size, embedding_dim, hidden_dim, n_layers, dropout=0.5)
if train_on_gpu:
    rnn.cuda()

# defining loss and optimization functions for training
optimizer = torch.optim.Adam(rnn.parameters(), lr=learning_rate)
criterion = nn.CrossEntropyLoss()

with active_session():
    trained_rnn = train_rnn(rnn, batch_size, optimizer, criterion, num_epochs, show_every_n_batches)

# saving the trained model
helper.save_model('./save/another_trained_rnn', trained_rnn)
print('Model Trained and Saved')


Training for 20 epoch(s)...
Epoch:    1/20    Loss: 5.037256064534187

Epoch:    1/20    Loss: 4.545198640108109

Epoch:    2/20    Loss: 4.3296525426386685

Epoch:    2/20    Loss: 4.228286460757255

Epoch:    3/20    Loss: 4.151110487126118

Epoch:    3/20    Loss: 4.095336674332619

Epoch:    4/20    Loss: 4.058167681140024

Epoch:    4/20    Loss: 4.0175933101177215

Epoch:    5/20    Loss: 3.9893080383485695

Epoch:    5/20    Loss: 3.963890566825867

Epoch:    6/20    Loss: 3.9482040604006543

Epoch:    6/20    Loss: 3.9297581963539123

Epoch:    7/20    Loss: 3.9092522415328435

Epoch:    7/20    Loss: 3.8850674464702606

Epoch:    8/20    Loss: 3.883890869862668

Epoch:    8/20    Loss: 3.865600531578064

Epoch:    9/20    Loss: 3.8507449864179364

Epoch:    9/20    Loss: 3.8452409970760346

Epoch:   10/20    Loss: 3.830766844241917

Epoch:   10/20    Loss: 3.823945849299431

Epoch:   11/20    Loss: 3.811235603842255

Epoch:   11/20    Loss: 3.8036593745946883

Epoch:   12/20  

/opt/conda/lib/python3.6/site-packages/torch/serialization.py:193: UserWarning: Couldn't retrieve source code for container of type RNN. It won't be checked for correctness upon loading.
  "type " + obj.__name__ + ". It won't be checked "


Model Trained and Saved


In [15]:
# Continue training
num_epochs = 5
learning_rate = 1e-4

with active_session():
    trained_rnn = train_rnn(rnn, batch_size, optimizer, criterion, num_epochs, show_every_n_batches)

# saving the trained model
helper.save_model('./save/another_trained_rnn', trained_rnn)
print('Model Trained and Saved')


Training for 5 epoch(s)...
Epoch:    1/5     Loss: 3.65164717400074

Epoch:    1/5     Loss: 3.675314612030983

Epoch:    2/5     Loss: 3.6710105270078532

Epoch:    2/5     Loss: 3.666993824481964

Epoch:    3/5     Loss: 3.6512115435833867

Epoch:    3/5     Loss: 3.6626744681596755

Epoch:    4/5     Loss: 3.6485013098984567

Epoch:    4/5     Loss: 3.6474142385721207

Epoch:    5/5     Loss: 3.63973519214455

Epoch:    5/5     Loss: 3.636299873948097

Model Trained and Saved


/opt/conda/lib/python3.6/site-packages/torch/serialization.py:193: UserWarning: Couldn't retrieve source code for container of type RNN. It won't be checked for correctness upon loading.
  "type " + obj.__name__ + ". It won't be checked "


In [16]:
# Train again
sequence_length = 15
batch_size = 150
embedding_dim = 200
hidden_dim = 250
train_loader = batch_data(int_text, sequence_length, batch_size)
num_epochs = 10
learning_rate = 1e-3
n_layers = 2
show_every_n_batches = 2500

# create model and move to gpu if available
rnn = RNN(vocab_size, output_size, embedding_dim, hidden_dim, n_layers, dropout=0.5)
if train_on_gpu:
    rnn.cuda()

# defining loss and optimization functions for training
optimizer = torch.optim.Adam(rnn.parameters(), lr=learning_rate)
criterion = nn.CrossEntropyLoss()

# training the model
trained_rnn = train_rnn(rnn, batch_size, optimizer, criterion, num_epochs, show_every_n_batches)

# saving the trained model
helper.save_model('./save/trained_rnn', trained_rnn)
print('Model Trained and Saved')

Training for 10 epoch(s)...
Epoch:    1/10    Loss: 4.886578718852997

Epoch:    1/10    Loss: 4.41216743478775

Epoch:    2/10    Loss: 4.119939510931611

Epoch:    2/10    Loss: 3.9242354477882384

Epoch:    3/10    Loss: 3.831624794961129

Epoch:    3/10    Loss: 3.725136310958862

Epoch:    4/10    Loss: 3.66002606619117

Epoch:    4/10    Loss: 3.5948093066215514

Epoch:    5/10    Loss: 3.5417292219085605

Epoch:    5/10    Loss: 3.4875552315711973

Epoch:    6/10    Loss: 3.442039058295641

Epoch:    6/10    Loss: 3.4036220861434936

Epoch:    7/10    Loss: 3.3705755169835814

Epoch:    7/10    Loss: 3.3330370136260985

Epoch:    8/10    Loss: 3.311113720568404

Epoch:    8/10    Loss: 3.2763917932510376

Epoch:    9/10    Loss: 3.261023654954349

Epoch:    9/10    Loss: 3.2284179419517516

Epoch:   10/10    Loss: 3.219365715357846

Epoch:   10/10    Loss: 3.1921128586769103



/opt/conda/lib/python3.6/site-packages/torch/serialization.py:193: UserWarning: Couldn't retrieve source code for container of type RNN. It won't be checked for correctness upon loading.
  "type " + obj.__name__ + ". It won't be checked "


Model Trained and Saved


### Question: How did you decide on your model hyperparameters? 
For example, did you try different sequence_lengths and find that one size made the model converge faster? What about your hidden_dim and n_layers; how did you decide on those?

**Answer:**

My first set of parameters were:
- Num of epochs = `3`
- Learning Rate = `1e-5`
- Sequence Length = `200`
- Batch Size = `10`
- Embedding Dimension = `15`
- Hidden Dimension = `15`
- Num of Layers = `2`
- Training Loss (approximate) = `9.8522`
- Still training in the first epoch, I discovered that the loss was decreasing but slowly.


My second set of parameters were:
- Num of epochs = `3`
- Learning Rate = `1e-3`
- Sequence Length = `200`
- Batch Size = `100`
- Embedding Dimension = `150`
- Hidden Dimension = `150`
- Num of Layers = `2`
- Training Loss (approximate) = `4.4616`

My third set of parameters were:
- Num of epochs = `2`
- Learning Rate = `3e-3`
- Sequence Length = `200`
- Batch Size = `100`
- Embedding Dimension = `150`
- Hidden Dimension = `150`
- Num of Layers = `2`
- Training Loss (approximate) = `4.3518`


A lot of other configurations gave me errors not lower than 3.6, and I still needed to reach lower `3.5`. And after several trials with diffierent model parameters, I decided to try much smaller values for the sequence length.
- Sequence Length = `15`
- Batch Size = `150`
- Learning Rate = `1e-3`
- Embedding Dimension = `200`
- Hidden Dimension = `250`
- I clipped the gradient to 5, to prevent the problem of exploding gradients, typical in deep recurrent neural networds.
- I disabled the dropout layer 
- I observed some improvement.
- Training Loss (average) = `3.1921`, which is below `3.5`

---
# Checkpoint

After running the above training cell, your model will be saved by name, `trained_rnn`, and if you save your notebook progress, **you can pause here and come back to this code at another time**. You can resume your progress by running the next cell, which will load in our word:id dictionaries _and_ load in your saved model by name!

In [17]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import torch
import helper
import problem_unittests as tests

_, vocab_to_int, int_to_vocab, token_dict = helper.load_preprocess()
# trained_rnn = helper.load_model('./save/trained_rnn')
trained_rnn = helper.load_model('./save/trained_rnn')

## Generate TV Script
With the network trained and saved, you'll use it to generate a new, "fake" Seinfeld TV script in this section.

### Generate Text
To generate the text, the network needs to start with a single word and repeat its predictions until it reaches a set length. You'll be using the `generate` function to do this. It takes a word id to start with, `prime_id`, and generates a set length of text, `predict_len`. Also note that it uses topk sampling to introduce some randomness in choosing the most likely next word, given an output set of word scores!

In [18]:
"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
import torch.nn.functional as F

def generate(rnn, prime_id, int_to_vocab, token_dict, pad_value, predict_len=100):
    """
    Generate text using the neural network
    :param decoder: The PyTorch Module that holds the trained neural network
    :param prime_id: The word id to start the first prediction
    :param int_to_vocab: Dict of word id keys to word values
    :param token_dict: Dict of puncuation tokens keys to puncuation values
    :param pad_value: The value used to pad a sequence
    :param predict_len: The length of text to generate
    :return: The generated text
    """
    rnn.eval()
    
    # create a sequence (batch_size=1) with the prime_id
    current_seq = np.full((1, sequence_length), pad_value)
    current_seq[-1][-1] = prime_id
    predicted = [int_to_vocab[prime_id]]
    
    for _ in range(predict_len):
        if train_on_gpu:
            current_seq = torch.LongTensor(current_seq).cuda()
        else:
            current_seq = torch.LongTensor(current_seq)
        
        # initialize the hidden state
        hidden = rnn.init_hidden(current_seq.size(0))
        
        # get the output of the rnn
        output, _ = rnn(current_seq, hidden)
        
        # get the next word probabilities
        p = F.softmax(output, dim=1).data
        if(train_on_gpu):
            p = p.cpu() # move to cpu
         
        # use top_k sampling to get the index of the next word
        top_k = 5
        p, top_i = p.topk(top_k)
        top_i = top_i.numpy().squeeze()
        
        # select the likely next word index with some element of randomness
        p = p.numpy().squeeze()
        word_i = np.random.choice(top_i, p=p/p.sum())
        
        # retrieve that word from the dictionary
        word = int_to_vocab[word_i]
        predicted.append(word)     
        
        # the generated word becomes the next "current sequence" and the cycle can continue
        current_seq = current_seq.cpu()
        current_seq = np.roll(current_seq, -1, 1)
        current_seq[-1][-1] = word_i
    
    gen_sentences = ' '.join(predicted)
    
    # Replace punctuation tokens
    for key, token in token_dict.items():
        ending = ' ' if key in ['\n', '(', '"'] else ''
        gen_sentences = gen_sentences.replace(' ' + token.lower(), key)
    gen_sentences = gen_sentences.replace('\n ', '\n')
    gen_sentences = gen_sentences.replace('( ', '(')
    
    # return all the sentences
    return gen_sentences

### Generate a New Script
It's time to generate the text. Set `gen_length` to the length of TV script you want to generate and set `prime_word` to one of the following to start the prediction:
- "jerry"
- "elaine"
- "george"
- "kramer"

You can set the prime word to _any word_ in our dictionary, but it's best to start with a name for generating a TV script. (You can also start with any other names you find in the original text file!)

In [24]:
# run the cell multiple times to get different results!
gen_length = 800 # modify the length to your preference
prime_word = 'kramer' # name for starting the script

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
pad_word = helper.SPECIAL_WORDS['PADDING']
generated_script = generate(trained_rnn, vocab_to_int[prime_word + ':'], int_to_vocab, token_dict, vocab_to_int[pad_word], gen_length)
print(generated_script)

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:42: UserWarning: RNN module weights are not part of single contiguous chunk of memory. This means they need to be compacted at every call, possibly greatly increasing memory usage. To compact weights again call flatten_parameters().


kramer: the whole time... the honorable judge county.

hoyt: the prosecution house]

hoyt: you were going to be polite.

hoyt: what?

george: i was just going to do this.

elaine: i don't know if i can see you.

george: what?

george: i think it was a good time, you can get that.

jerry: oh, no, no. no.

george: oh my god.

hoyt: you know what? you can go on the plane, and then i was standing up. i can't do this.

hoyt: the moops

boy: oh, i was not going to do.

hoyt: what?

george: what about that woman?

jerry: yeah, but, you didn't want to get it?

george: oh. oh, no no.

jerry: you want to get it?

george: i think it's the only ones that can be.

george: what is that noise?

jerry: no, not. i mean, i think they had to go.

george: well, the mets are going to be the maximum jet for the lipo.

hoyt: so, what is it?

george: yeah, i think i could get that jet in this contest.

hoyt: i think i was in latham.

george: what?

hoyt: i don't know.

jerry: oh, hi.

george: oh, i don't know

#### Save your favorite scripts

Once you have a script that you like (or find interesting), save it to a text file!

In [25]:
# save script to a text file
f =  open("generated_script_1.txt","w")
f.write(generated_script)
f.close()

# The TV Script is Not Perfect
It's ok if the TV script doesn't make perfect sense. It should look like alternating lines of dialogue, here is one such example of a few generated lines.

### Example generated script

>jerry: what about me?
>
>jerry: i don't have to wait.
>
>kramer:(to the sales table)
>
>elaine:(to jerry) hey, look at this, i'm a good doctor.
>
>newman:(to elaine) you think i have no idea of this...
>
>elaine: oh, you better take the phone, and he was a little nervous.
>
>kramer:(to the phone) hey, hey, jerry, i don't want to be a little bit.(to kramer and jerry) you can't.
>
>jerry: oh, yeah. i don't even know, i know.
>
>jerry:(to the phone) oh, i know.
>
>kramer:(laughing) you know...(to jerry) you don't know.

You can see that there are multiple characters that say (somewhat) complete sentences, but it doesn't have to be perfect! It takes quite a while to get good results, and often, you'll have to use a smaller vocabulary (and discard uncommon words), or get more data.  The Seinfeld dataset is about 3.4 MB, which is big enough for our purposes; for script generation you'll want more than 1 MB of text, generally. 

# Submitting This Project
When submitting this project, make sure to run all the cells before saving the notebook. Save the notebook file as "dlnd_tv_script_generation.ipynb" and save another copy as an HTML file by clicking "File" -> "Download as.."->"html". Include the "helper.py" and "problem_unittests.py" files in your submission. Once you download these files, compress them into one zip file for submission.